### PR TITLE
boringssl --> OpenSSL 3.2

### PR DESCRIPTION
**Pull request by rschu1ze**
Like #56398, but OpenSSL 3.2. Made a new PR so that it doesn't become too messy.

Background:
-  OpenSSL seems to become more popular (compared to boringssl) with the improvements done after [Heartbleed](https://heartbleed.com/). As both codebases diverge, ClickHouse had to introduce hacks to force submodules to link against boringssl when upstream supported only OpenSSL. These hacks will become obsolete with this PR.
- Should enable FIPS 140-3 with TLS 1.3 (#34238) *in the long run*. BoringSSL doesn't seem to have a TLS 1.3 FIPS compliant version (aka. "boringcrypto"). It is a bit tricky though... as per Feb 2024, OpenSSL 3.0.8 and 3.0.9 are FIPS-140-2 validated ([announcement](https://www.openssl.org/blog/blog/2024/01/23/fips-309/)), OpenSSL 3.1.2 is under FIPS-140-3 validation ([announcement)](https://www.openssl.org/blog/blog/2023/10/12/osslfips-timeline/index.html), and OpenSSL 3.2 appears too new so no validation was seemingly started yet. Also see #34238.
- OpenSSL (no specific version requirement) is a prerequisite for upgrading Azure SDK from 1.7 to 1.8 ([see here](https://github.com/ClickHouse/ClickHouse/pull/57881#issuecomment-1867855732))
- We specifically need to go with OpenSSL 3.2 instead of 3.0 or 3.1 because of this ~mess~ [constraint](https://github.com/ClickHouse/ClickHouse/pull/56398#issuecomment-1937766144), i.e. all data stored with `AES_128_GCM_SIV` and `AES_256_GCM_SIV` codecs would no longer be decrypt-able.

Details:
- Prior to this PR, ClickHouse linked against boringssl pulled in as submodule. An unsupported alternative was to link OpenSSL statically (`ENABLE_OPENSSL`) via a submodule or dynamically (`ENABLE_OPENSSL AND ENABLE_OPENSSL_DYNAMIC`) against the OS-provided OpenSSL, usually a FIPS-certified one.
- https://github.com/ClickHouse/ClickHouse/pull/59870/commits/e85adb95292cc38bd108f38fd0af571d4e38752a removes the boringssl submodule and https://github.com/ClickHouse/ClickHouse/pull/59870/commits/8b259d5bf0b3888c36b8309597cee86a5943bea3 switches the build to the statically linked OpenSSL version.
- That broke the [sparse checkout](https://clickhouse.com/docs/en/development/developer-instruction#cloning-a-repository-to-your-development-machine). Fixed with https://github.com/ClickHouse/ClickHouse/pull/59870/commits/096e35874fd911a2820002b702e7f37b56247489
- Next replace the OpenSSL submodule (pointing to upstream) by a forked one (https://github.com/ClickHouse/ClickHouse/pull/59870/commits/bb906e04e4f54b1e47cee36046381580b915e69b, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/b61ee06b029df1a5071c6baa5ac1446e5ffd2635)
- The previous OpenSSL submodule was 3.0, the new one is 3.2. This breaks the build for all platforms. To fix it, one needs to create new CMake descriptions from OpenSSL's makefile (for detailed steps, see the beginning of contrib/openssl-cmake/CMakeLists.txt + https://github.com/ClickHouse/ClickHouse/pull/43991#discussion_r1399841779).
- https://github.com/ClickHouse/ClickHouse/pull/59870/commits/124b8820b69408f0cf5d7af75ca9e9ba7afa9d6f fixes x86, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/e13bd77da889365a9fb8f655b7926b8339613d0d fixes ARM, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/f00b73cf6a4acc5678326b85d9df1a10d9f24803 fixes PPC, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/0b5031819a4fc5eee26d84d209327d66a1b837f2 fixes macOS x86 and ARM, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/ea4fdb38189cb62696aa1b04d481b6b2412ada9d fixes RISC-V, and https://github.com/ClickHouse/ClickHouse/pull/59870/commits/b6528e65764269e039f0af6b455404286b6f82e1 rage-disables s390x because https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1959910652 and https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1961451875 ... I could not figure out how to make the build work (collateral damage, sorry IBM). We should enable it back later, see https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1978990248. Lots of trash accumulated while re-enabling the platforms, the garbage is removed with https://github.com/ClickHouse/ClickHouse/pull/59870/commits/9499484b330f2b76d6d57b2fbbae95b1529d76b2.
- Some tests gave different error codes and could be fixed easily: https://github.com/ClickHouse/ClickHouse/pull/59870/commits/fa74ae4b541c58c934b27a55ccc8e266254edf9b, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/ee588b6860be1e73a9dec34d99780ddef452899a, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/d93d4de9408a1c32a1eaa5d0f84eff1b973cd766
- Glibc compat check in CI complained OpenSSL 3.2 needs a too-new glibc, fixed with https://github.com/ClickHouse/ClickHouse/pull/59870/commits/cb5db8c54acdc79b2087104611393763395d4a2a and https://github.com/ClickHouse/ClickHouse/pull/59870/commits/cdeef25a4e2da63a8cf2061c9ec55bc883b75e2c
- Tests for CompressionCodecEncrypted failed with the most obscure errors. I remember debugging for hour, reading scattered tutorial fragments before finding out that GCM_SIV requires to call the API in a slightly different way. Fixed with https://github.com/ClickHouse/ClickHouse/pull/59870/commits/7e5432f477f4adfc9e67754a3d8dc091632bf1ca. This commit also provides backward compatibility for s390x. https://github.com/ClickHouse/ClickHouse/pull/59870/commits/dd73ccff52dad62ecb37fea5a989b0ab0e226db9 is a necessary refactoring to make that all work.
- Integration test test_dictionaries_all_layouts_separate_sources failed with obscure "unexpected EOF" errors. By pure luck, I found https://github.com/openssl/openssl/discussions/22690#discussioncomment-7534909 and could fix that: https://github.com/ClickHouse/ClickHouse/pull/59870/commits/55d35d8d60e0ba397b787a553cdce81df0be5ff0
- More fixes for msan failures: https://github.com/ClickHouse/ClickHouse/pull/59870/commits/08c4600014a19f141097670122d90ff328011129 https://github.com/ClickHouse/ClickHouse/pull/59870/commits/979e06c7679b04e5c6eb0d5e1c7fa57c39282d77
- And fixes for tsan: https://github.com/ClickHouse/ClickHouse/pull/59870/commits/bdaace3fd5ef5cc4f0b214dabd5054e38f5cefd9 This disables atomics in OpenSSL but I don't think that's a performance problem.
- And fixes for leaksan: https://github.com/ClickHouse/ClickHouse/pull/59870/commits/e06a9c3df20da51e1001396408ddedd004d884ae, https://github.com/ClickHouse/ClickHouse/pull/59870/commits/164cd580a295dbf6a83716ec58974f5d90ca899b
- 01747_system_session_log_long was fixed by https://github.com/ClickHouse/ClickHouse/pull/59870/commits/5d37a459090334f3dadc04056d4236cc8d7dbb3c. The root cause for this was a bug in libpq (PostgreSQL, [here](https://github.com/postgres/postgres/commit/c82207a548db47623a2bfa2447babdaa630302b9)) which was fixed upstream in the meantime. Backported the fix into ClickHouse's stone age copy of libpq.
- Finally, there were some data races detected by tsan in integration tests. These point to problems with session caching in OpenSSL. Force-disabled it for now with https://github.com/ClickHouse/ClickHouse/pull/59870/commits/d30b48f54e42ec5aeb1bc27c937adb8efdb8108d and will check further.

### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 125a97e901c7d7b384a07c8042310ca60cf13afc with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59870/125a97e901c7d7b384a07c8042310ca60cf13afc/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59870/125a97e901c7d7b384a07c8042310ca60cf13afc/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by UnamedRus** [comment]
One of the reason of going to boringSSL was performance speedup (30x) for encryption functions https://github.com/ClickHouse/ClickHouse/pull/11844#issuecomment-753657434

**Comment by rschu1ze** [comment]
The azure submodule worked with boringssl only by chance, and the 1.8 upgrade no longer builds at all with boringssl. That's a blocker.

I assume that the performance measurements were done against OpenSSL 1.x, perhaps runtimes improved in the meantime.

**Comment by rschu1ze** [comment]
Status update:

- Applying https://github.com/ClickHouse/ClickHouse/pull/56398/commits/b442b2254925ca2b3beb2e027b982429ced6f417 to this PR and using cipher "AES-128-GCM-SIV" instead of "AES-128-GCM" works. This will resolve https://github.com/ClickHouse/ClickHouse/pull/56398#issuecomment-1937766144

- Re-generating CMake build descriptions after the upgrade was daunting (to put it nicely). I am tempted to do the same exercise only for ARM, and use much simpler build descriptions for all other platforms (e.g. no assembly).

**Comment by rschu1ze** [comment]
`readelf` dies for me when running locally. Not clear how to reproduce the existing (which should be fixed by https://github.com/ClickHouse/ClickHouse/pull/56398/commits/9eb2ecfe02e2be6447502f3cad9d316096190e5e) and the new problem in the compatibility check.

**Comment by rschu1ze** [comment]
> Applying https://github.com/ClickHouse/ClickHouse/commit/b442b2254925ca2b3beb2e027b982429ced6f417 to this PR and using cipher "AES-128-GCM-SIV" instead of "AES-128-GCM" produces the expected result. This will resolve https://github.com/ClickHouse/ClickHouse/pull/56398#issuecomment-1937766144

Not true. Encryption works and produces the right result, decryption throws an error with unhelpful error text `error:00000000:lib(0)::reason(0)`. No clue yet why, probably because we are now in "SIV Mode" ([manpage](https://www.openssl.org/docs/manmaster/man3/EVP_CIPHER_iv_length.html)) and the calls to OpenSSL need to be changed in subtle ways.

**Comment by rschu1ze** [comment]
[Aforementioned](https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1939478585) manpage states in section `SIV Mode`:

```
EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_AEAD_SET_TAG, taglen, tag);
Sets the expected tag (the Synthetic Initialization Vector) to taglen bytes from tag.
This call is only legal when decrypting data and must be made before any data is
processed (e.g. before any EVP_DecryptUpdate() calls). For SIV mode the taglen must be
16.
```

Running `EVP_CIPHER_CTX_ctrl()` *before* `EVP_DecryptUpdate()` did the trick, decompression now works.
(It is still terrifying to not get a proper error message for such types of API misuse.)

**Comment by rschu1ze** [comment]
> Encryption works and produces the right result, decryption throws an error with generic error message error:00000000:lib(0)::reason(0) (ironically meaning "no error" when you google this error). 

I am pretty sure that the error handling in our existing code calling OpenSSL is wrong (this leads to different actual and expected error codes for a negative test in integration test `test_replicated_merge_tree_encryption_codec`). Basically, OpenSSL stores errors in a queue, so when the 5th call fails, one needs to loop through and ignore four previous errors using `ERR_get_error()` OR clear the queue before each call using `ERR_clear_error()`. We previously only called `ERR_get_error()` which gets the earliest error. The code was only exercised on s390/x and thus untested.

On the hand, doing `ERR_clear_error()` before each call to OpenSSL functions also doesn't produce better error codes, something else seems wrong in OpenSSL.

**Comment by rschu1ze** [comment]
Two more integration tests with wrong expected error coded fixed, working now on getting all non-x86 build working again.

**Comment by rschu1ze** [comment]
> trying now to restore the non-x86 builds.

For ARM, I tried the cross-compiling instructions mentioned here https://github.com/ClickHouse/ClickHouse/pull/43991#discussion_r1400857757. `./Configure` generates the makefile but then mixes compiler calls (which fail) with header-generating steps, such that not all headers needed for copy into the platform-specific build directories exist. I guess the compiler errors are because I don't have the necessary toolchain set up. Need to dig deeper ...

For Arm, I'll also try on my EC2 ARM instance.

**Comment by rschu1ze** [comment]
ARM build are fixed now, checking now why integration test `test_dictionaries_all_layouts_separate_sources` fails.

**Comment by rschu1ze** [comment]
Need to work on other stuff with a deadline until Friday.

**Comment by rschu1ze** [comment]
PPC builds should work now.

Tried to fix RISC-V builds locally but when I try to cross-compile ClickHouse, I run into:

```
/usr/bin/riscv64-linux-gnu-ld.bfd: /home/ubuntu/repo/ch2/cmake/linux/../../contrib/sysroot/linux-riscv64/lib/riscv64-linux-gnu/librt.a(timer_create.o): undefined reference to symbol '__stack_chk_guard@@GLIBC_2.27
/usr/bin/riscv64-linux-gnu-ld.bfd: /home/ubuntu/repo/ch2/cmake/linux/../../contrib/sysroot/linux-riscv64/lib/riscv64-linux-gnu/./ld-linux-riscv64-lp64d.so.1: error adding symbols: DSO missing from command line
lang++-17: error: linker command failed with exit code 1 (use -v to see invocation)
```

plus the build is ultra slow locally.

Also tried to fix s390x builds but my attempt to cross-compile locally (without the boring-to-openssl transition) failed with weird linker errors.. Additionally, a standalone openssl (which is needed to generate platform-specifc files) refuses to compile on LE systems (like x86) for BE systems (s390x). Luckily, it generated at least a few critical files, other ones (which seem very similar across platforms) I could copy from other platforms.

**Comment by rschu1ze** [comment]
Yes, RISC-V works.

Darwin ARM builds now compile and link successfully locally (with actual `darwin-aarch64` platform directory unlike what we had in https://github.com/ClickHouse/ClickHouse/pull/56398/commits). However, when I start clickhouse-server on my Mac I get

```
~/repo/ClickHouse/build/programs (be-less-boring-32 *>) $ chs
dyld[93091]: dyld cache '(null)' not loaded: syscall to map cache into shared region failed
dyld[93091]: Library not loaded: /usr/lib/libiconv.2.dylib
  Referenced from: <C296BB2F-0441-3025-B1C6-D1AA4E4243B3> /Users/robert/repo/ClickHouse/build/programs/clickhouse
  Reason: tried: '/usr/lib/libiconv.2.dylib' (no such file), '/System/Volumes/Preboot/Cryptexes/OS/usr/lib/libiconv.2.dylib' (no such file), '/usr/lib/libiconv.2.dylib' (no such file, no
dyld cache)
Abort trap: 6
```

which does not look related but I'll make a `master` build to double-check.

I also generated similar platform files for darwin-x86_64 and there is a chance it works, but I did not test locally yet (cross-compiling for Mac is troublesome locally)

**Comment by rschu1ze** [comment]
> Regarding s390x ([here](https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1959910652)):
> 
>     * I followed the [documented steps](https://clickhouse.com/docs/en/development/build-cross-s390x) with a `master` build (w/o this PR) and linking fails with `clang++-17: error: invalid linker name in argument '-fuse-ld=mold'`. I then installed `mold` (`apt install mold`) and linking fails with `mold: unknown -m argument: elf64_s390`. What am I doing wrong here? @bkuschel @larryluogit  Perhaps you have an idea or updated build instructions?

Temporarily disabled s390/x.

**Comment by Algunenano** [comment]
> I followed the [documented steps](https://clickhouse.com/docs/en/development/build-cross-s390x) with a master build (w/o this PR) and linking fails with clang++-17: error: invalid linker name in argument '-fuse-ld=mold'. I then installed mold (apt install mold) and linking fails with mold: unknown -m argument: elf64_s390. What am I doing wrong here? @bkuschel @larryluogit Perhaps you have an idea or updated build instructions?

@yakov-olkhovskiy is the one that hardcoded mold

**Comment by rschu1ze** [comment]
I guess you mean the mold linker was hardcoded here: https://github.com/ClickHouse/ClickHouse/pull/59870/files#diff-cc10102f8b1db393176bb4ef3536718e0d91ca8b6c086e088953b0d951732c9bR82
If I have time, I can check if I get s390x working with that, but there is a chance that it will stay disabled in CI until this PR is merged.

**Comment by yakov-olkhovskiy** [comment]
Linux s390x build was added here:
https://github.com/ClickHouse/ClickHouse/pull/53181
mold was the only linker working there...

**Comment by rschu1ze** [comment]
> Regarding compat check:
> 
>     * About [*wip* boringssl --> OpenSSL 3.2 #59870 (comment)](https://github.com/ClickHouse/ClickHouse/pull/59870#issuecomment-1939445522)): It turned out 1. I was wrongly using `BuilderDebRelease` (which uses musl) as template for local build to run the compat check. It should have been `BuilderBinRelease` 2. With this, `readelf` from `binutils` still died, but installing `elfutils` and using `eu-readutils` instead worked.
> 
>     * Picking [9eb2ecf](https://github.com/ClickHouse/ClickHouse/commit/9eb2ecfe02e2be6447502f3cad9d316096190e5e) into this PR made the ARM compatibility check green. Interestingly, x86 one is still red: https://s3.amazonaws.com/clickhouse-test-reports/59870/0f3103f494b8b7ccd7da62ff9d3cc23932c42a7b/compatibility_check__amd64_.html with two other too-new symbols `sendmmsg@GLIBC_2.14` and `recvmmsg@GLIBC_2.12`.
> 
>     * Not clear yet what causes this, one candidate are new `BIO_s_dgram_pair()`/ `BIO_s_dgram_mem()` calls added with OpenSSL 3.2. (see their [changelog](https://github.com/openssl/openssl/blob/master/CHANGES.md)).

Fixed now (--> https://github.com/ClickHouse/ClickHouse/pull/59870/commits/fb00fa64d58ca452ef2ca07b56f9b61fa115a164). Interestingly, the same problem with old glibc-s appeared earlier: https://github.com/openssl/openssl/pull/22036

**Comment by rschu1ze** [comment]
> [...], checking now why integration test `test_dictionaries_all_layouts_separate_sources` fails.

This one was tricky but I am pretty sure I fixed it. The test failed with error

```
<Information> ReadWriteBufferFromHTTP: Failed to make request to `https://localhost:8889/`. Error: SSL Exception: error:0A000126:SSL routines::unexpected eof while reading. Failed at try 9/10. Will retry with current backoff wait is 10000/10000 ms.
```

After some googling, this seemed like a behavior change from OpenSSL 1.x to 3.x: https://github.com/openssl/openssl/discussions/22690#discussioncomment-7534909. `SSL_CTX_set_options(_pSSLContext, SSL_OP_IGNORE_UNEXPECTED_EOF);` goes back to the previous behavior.

**Comment by Slach** [comment]
I hope switch to openssl 3.x will no affect with performance degradation look details https://github.com/openssl/openssl/issues/17064

**Comment by rschu1ze** [comment]
Tried for two hours to reproduce [this failure](https://s3.amazonaws.com/clickhouse-test-reports/59870/58c3aa77d9c4a27439a655065a721f91a89fc451/stateless_tests__tsan__[3_5].html) in 01683_text_log_deadlock which uses `clickhouse-benchmark --secure`. Did not find a way to make this command run even on a build without this PR, I am constantly getting `CERTIFICATE_VERIFY_FAILED` errors :-(

File: tests/queries/0_stateless/01683_text_log_deadlock.sh [hunk 1]
```diff
@@ -1,5 +1,5 @@
 #!/usr/bin/env bash
-# Tags: deadlock
+# Tags: deadlock, no-tsan
```
Comment (hanfei1991) [inline tests/queries/0_stateless/01683_text_log_deadlock.sh:2]: It seems we have supressed tsan issues here https://github.com/ClickHouse/ClickHouse/pull/59870/files#diff-3cdbc9c2a0c3de2ee8e6fd929808e956ed2b0a27d183a9040a89863da09d272fR333 Why do we tag `no-tsan` in the tests?

File: docs/en/development/build-cross-s390x.md [hunk 2]
```diff
@@ -10,11 +10,11 @@ As of writing (2023/3/10) building for s390x considered to be experimental. Not
 
 ## Building
 
-As s390x does not support boringssl, it uses OpenSSL and has two related build options. 
+s390x has two OpenSSL-related build options.
```
Comment (hanfei1991) [inline docs/en/development/build-cross-s390x.md:13]: According to the comments, s390x `refused to build in the migration to OpenSSL`. Do we need to mention it here?

**Review by hanfei1991 — COMMENTED** [review]
the rest LGTM. Awesome work!

File: docs/en/development/build-cross-s390x.md (hunk 2 above)
Comment (rschu1ze) [reply docs/en/development/build-cross-s390x.md:13]: Done, thanks.

File: tests/queries/0_stateless/01683_text_log_deadlock.sh (hunk 1 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/01683_text_log_deadlock.sh]: I remember I could not get this test and 01393_benchmark_secure_port.sh to run locally. I just saw that they consistently failed in tsan builds. The log files from CI were also not helpful ... so I disabled them for now :-( Not great, but I'll check as a follow up.

**Comment by Algunenano** [comment]
FYI, I'm seeing tons of errors when building from scratch:

```
[271/13662] Generating ../openssl/crypto/bn/rsaz-3k-avx512.s
sh: 0: Illegal option --
sh: 0: Illegal option - 
[272/13662] Generating ../openssl/crypto/bn/rsaz-2k-avx512.s
sh: 0: Illegal option --
sh: 0: Illegal option - 
[273/13662] Generating ../openssl/crypto/aes/aesni-sha256-x86_64.s
sh: 0: Illegal option --
sh: 0: Illegal option - 
[274/13662] Generating ../openssl/crypto/bn/rsaz-4k-avx512.s
sh: 0: Illegal option -
```

Do we need a specific version of perl? Do we even need to generate those files instead of pushing them directly to the submodule?

**Comment by Algunenano** [comment]
It seems it needs CC to be declared externally:
```
perl -d:Trace /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl /mnt/ch/ClickHouse/build_public/contrib/openssl/crypto/bn/rsaz-2k-avx512.s
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:31: $output = $#ARGV >= 0 && $ARGV[$#ARGV] =~ m|\.\w+$| ? pop : undef;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:32: $flavour = $#ARGV >= 0 && $ARGV[0] !~ m|\.| ? shift : undef;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:34: $win64=0; $win64=1 if ($flavour =~ /[nm]asm|mingw64/ || $output =~ /\.asm$/);
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:34: $win64=0; $win64=1 if ($flavour =~ /[nm]asm|mingw64/ || $output =~ /\.asm$/);
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:35: $avx512ifma=0;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:37: $0 =~ m/(.*[\/\\])[^\/\\]+$/; $dir=$1;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:37: $0 =~ m/(.*[\/\\])[^\/\\]+$/; $dir=$1;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:38: ( $xlate="${dir}x86_64-xlate.pl" and -f $xlate ) or
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:42: if (`$ENV{CC} -Wa,-v -c -o /dev/null -x assembler /dev/null 2>&1`
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:47: if (!$avx512ifma && $win64 && ($flavour =~ /nasm/ || $ENV{ASM} =~ /nasm/) &&
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:52: if (!$avx512ifma && `$ENV{CC} -v 2>&1`
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:65: open OUT,"| \"$^X\" \"$xlate\" $flavour \"$output\""
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:67: *STDOUT=*OUT;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:69: if ($avx512ifma>0) {{{
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:718: }}} else {{{                # fallback for old assembler
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:719: $code.=<<___;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:719: $code.=<<___;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:742: $code =~ s/\`([^\`]*)\`/eval $1/gem;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:743: print $code;
>> /mnt/ch/ClickHouse/contrib/openssl/crypto/bn/asm/rsaz-2k-avx512.pl:744: close STDOUT or die "error closing STDOUT: $!";
sh: 0: Illegal option --
sh: 0: Illegal option -
```

I guess cmake should take care of that

**Comment by HarryLeeIBM** [comment]
Has anybody tried to build the latest with `cmake .. -DENABLE_OPENSSL_DYNAMIC=1`? It gives error:
```
/usr/bin/x86_64-linux-gnu-ld: /usr/bin/x86_64-linux-gnu-ld: DWARF error: invalid or unhandled FORM value: 0x25
contrib/openssl-cmake/CMakeFiles/crypto.dir/__/openssl/engines/e_dasync.c.o: in function `v_check':
e_dasync.c:(.text.v_check+0x0): multiple definition of `v_check'; /usr/bin/x86_64-linux-gnu-ld: DWARF error: invalid or unhandled FORM value: 0x25
contrib/openssl-cmake/CMakeFiles/crypto.dir/__/openssl/engines/e_capi.c.o:e_capi.c:(.text.v_check+0x0): first defined here
```

**Comment by HarryLeeIBM** [comment]
Regarding CH build for openssl dynamic linking issue, a PR has been raised:[62888](https://github.com/ClickHouse/ClickHouse/pull/62888)

**Pull request merged**